src/number_theory/cyclotomic/primitive_roots.lean

```diff
@@ -193,6 +202,154 @@ begin
+  (hζ: is_primitive_root ζ ↑(p ^ (k + 1))) [hpri : fact (p : ℕ).prime]
+  [is_cyclotomic_extension {p ^ (k + 1)} K L]
+  (hirr : irreducible (cyclotomic (↑(p ^ (k + 1)) : ℕ) K)) (h : p ≠ 2) :
+  norm K (ζ - 1) = p :=
+begin
+  haveI : ne_zero ((↑(p ^ (k + 1)) : ℕ) : K),
+  { refine ⟨λ hzero, _⟩,
+    rw [pow_coe] at hzero,
+    simpa [ne_zero.ne ((p : ℕ) : K)] using hzero },
+  have : 2 < p ^ (k + 1),
+  { rw [← coe_lt_coe, pow_coe, pnat.coe_bit0, one_coe],
+    calc 2 < (p : ℕ) : lt_of_le_of_ne hpri.1.two_le (by contrapose! h; exact coe_injective h.symm)
+      ...  = (p : ℕ) ^ 1 : (pow_one _).symm
+      ...  ≤ (p : ℕ) ^ (k + 1) : pow_le_pow (nat.prime.pos hpri.out) le_add_self },
+  simp [sub_one_norm_eq_eval_cyclotomic hζ this hirr]
+end
+
+/-- If `irreducible (cyclotomic p K)` (in particular for `K = ℚ`) and `p` is an odd prime,
+then the norm of `ζ - 1` is `p`. -/
+lemma prime_ne_two.sub_one_norm {p : ℕ+} [ne_zero ((p : ℕ) : K)] [hpri : fact (p : ℕ).prime]
+  [hcyc : is_cyclotomic_extension {p} K L] (hζ: is_primitive_root ζ p)
+  (hirr : irreducible (cyclotomic p K)) (h : p ≠ 2) :
+  norm K (ζ - 1) = p :=
+begin
+  replace hirr : irreducible (cyclotomic (↑(p ^ (0 + 1)) : ℕ) K) := by simp [hirr],
+  replace hζ : is_primitive_root ζ (↑(p ^ (0 + 1)) : ℕ) := by simp [hζ],
+  haveI : ne_zero ((↑(p ^ (0 + 1)) : ℕ) : K) := ⟨by simp [ne_zero.ne ((p : ℕ) : K)]⟩,
+  haveI : is_cyclotomic_extension {p ^ (0 + 1)} K L := by simp [hcyc],
+  simpa using prime_ne_two_pow.sub_one_norm 0 hζ hirr h
+end
+
+/-- If `irreducible (cyclotomic (2 ^ k) K)` (in particular for `K = ℚ`) and `k` is at least `2`,
+then the norm of `ζ - 1` is `2`. -/
+lemma two_pow.sub_one_norm [ne_zero (2 : K)] {k : ℕ} (hk : 2 ≤ k)
+  [is_cyclotomic_extension {2 ^ k} K L] (hζ: is_primitive_root ζ (2 ^ k))
+  (hirr : irreducible (cyclotomic (2 ^ k) K)) :
+  norm K (ζ - 1) = 2 :=
+begin
+  haveI : ne_zero (((2 ^ k : ℕ+) : ℕ) : K),
+  { refine ⟨λ hzero, _⟩,
+    rw [pow_coe, pnat.coe_bit0, one_coe, cast_pow, cast_bit0, cast_one,
+      pow_eq_zero_iff (lt_of_lt_of_le zero_lt_two hk)] at hzero,
+    exact (ne_zero.ne (2 : K)) hzero,
+    apply_instance },
+  have : 2 < (2 ^ k : ℕ+),
+  { simp only [← coe_lt_coe, pnat.coe_bit0, one_coe, pow_coe],
+    nth_rewrite 0 [← pow_one 2],
+    exact pow_lt_pow one_lt_two (lt_of_lt_of_le one_lt_two hk) },
+  replace hirr : irreducible (cyclotomic (2 ^ k : ℕ+) K) := by simp [hirr],
+  replace hζ : is_primitive_root ζ (2 ^ k : ℕ+) := by simp [hζ],
+  obtain ⟨k₁, hk₁⟩ := exists_eq_succ_of_ne_zero ((lt_of_lt_of_le zero_lt_two hk).ne.symm),
+  simpa [hk₁] using sub_one_norm_eq_eval_cyclotomic hζ this hirr,
+end
 
 end is_primitive_root
+
+namespace is_cyclotomic_extension
+
+open is_primitive_root
+
+/-- If `K` is linearly ordered (in particular for `K = ℚ`), the norm of `zeta n K L` is `1`
+if `n` is odd. -/
+lemma norm_zeta_eq_one (K : Type u) [linear_ordered_field K] (L : Type v) [field L] [algebra K L]
```

Comment:
I don't really see why we need all the lemmas specifically for `zeta`, `(zeta_primitive_root ...).` should be fairly natural in the proof

Reply:
I would say it's reasonable to have them, since we have an explicit definition for `zeta`, but I don't have any strong opinion.

Reply:
The `haveI` in the proof below suggests to me that there will be enough of a hassle to not make inlining worth it. So I vote for keeping them.